src/rm_topk.c

```diff
@@ -0,0 +1,245 @@
+//#include <math.h>    // ceil, log10f
+//#include <strings.h> // strncasecmp
+
+#include "version.h"
+#include "rmutil/util.h"
+
+#include "topk.h"
+#include "rm_topk.h"
+
+#define INNER_ERROR(x)                                                                             \
+    RedisModule_ReplyWithError(ctx, x);                                                            \
+    return REDISMODULE_ERR;
+
+RedisModuleType *TopKType;
+
+static int GetTopKKey(RedisModuleCtx *ctx, RedisModuleString *keyName, TopK **topk, int mode) {
+    // All using this function should call RedisModule_AutoMemory to prevent memory leak
+    RedisModuleKey *key = RedisModule_OpenKey(ctx, keyName, mode);
+    if (RedisModule_KeyType(key) == REDISMODULE_KEYTYPE_EMPTY) {
+        INNER_ERROR("TopK: key does not exist");
+    } else if (RedisModule_ModuleTypeGetType(key) != TopKType) {
+        INNER_ERROR(REDISMODULE_ERRORMSG_WRONGTYPE);
+    }
+    *topk = RedisModule_ModuleTypeGetValue(key);
+    RedisModule_CloseKey(key); //TODO: check if this works
+    return REDISMODULE_OK;
+}
+
+
+static int createTopK(RedisModuleCtx *ctx, RedisModuleString **argv, int argc, TopK **topk) {
+    long long k, width, depth;
+    double decay;
+    if ((RedisModule_StringToLongLong(argv[2], &k) != REDISMODULE_OK) || k < 1) {
+        INNER_ERROR("TopK: invalid k");
+    }
+    if ((RedisModule_StringToLongLong(argv[3], &width) != REDISMODULE_OK) || width < 1) {
+        INNER_ERROR("TopK: invalid width");
+    }
+    if ((RedisModule_StringToLongLong(argv[4], &depth) != REDISMODULE_OK) || depth < 1) {
+        INNER_ERROR("TopK: invalid depth");
+    }    
+    if ((RedisModule_StringToDouble(argv[5], &decay) != REDISMODULE_OK) ||
+                                    (decay <= 0 || decay >= 1)) {
+        INNER_ERROR("TopK: invalid decay value. must be '> 1' & '< 0'");
+    }
+
+    *topk = TopK_Create(k, width, depth, decay);
```

Comment:
The level of strictness you want to maintain about this is up to you, but your long longs are implicitly getting demoted to unsigned ints here. Which is likely fine, but good to be aware of!

Reply:
I'm using long long b/c it is converted from redisStrings and that is my only option for an unsigned variable.

Reply:
Yeah, the module API is unfortunately a bit lacking with regard to comprehensive type handling. The safer alternative would be checking like:
```
    if ((RedisModule_StringToLongLong(argv[3], &width) != REDISMODULE_OK) || width < 1 || width > UINT_MAX) {
        INNER_ERROR("TopK: invalid width");
```
But that's definitely a standard of safety we don't always hold ourselves to.

No changes needed, I just wanted to highlight this area!